src/rgw/rgw_rados.cc

```diff
@@ -5723,12 +5779,107 @@ int RGWRados::swift_versioning_copy(RGWBucketInfo& bucket_info, RGWRados::Object
                NULL, /* void (*progress_cb)(off_t, void *) */
                NULL); /* void *progress_data */
   if (r == -ECANCELED || r == -ENOENT) { /* has already been overwritten, meaning another rgw process already copied it out */
+  ldout(cct, 20) << "++++ ECANC   --- SWIFT VERSION COPY BEGIN ++++ LINE: " << __LINE__ << dendl;
     return 0;
   }
 
   return r;
 }
 
+int RGWRados::swift_versioning_restore(RGWObjectCtx& obj_ctx,
+                                       const rgw_user& user,
+                                       RGWBucketInfo& bucket_info,
+                                       rgw_obj& obj,
+                                       bool& restored)             /* out */
+{
+  if (swift_versioning_enabled(bucket_info) == false) {
+    return 0;
+  }
+
+  /* Bucket info of the bucket that stores previous versions of our object. */
+  RGWBucketInfo archive_binfo;
+
+  int ret = get_bucket_info(obj_ctx, bucket_info.bucket.tenant,
```

Comment:
@rzarzynski I wonder if we need to keep also the target bucket id, and do a get_bucket_instance_info() here.. probably not though